pkg/compact/downsample/streamed_block_writer.go

```diff
@@ -0,0 +1,272 @@
+package downsample
+
+import (
+	"io"
+	"path/filepath"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/thanos/pkg/block"
+	"github.com/improbable-eng/thanos/pkg/block/metadata"
+	"github.com/improbable-eng/thanos/pkg/runutil"
+	"github.com/pkg/errors"
+	"github.com/prometheus/tsdb"
+	"github.com/prometheus/tsdb/chunks"
+	"github.com/prometheus/tsdb/fileutil"
+	"github.com/prometheus/tsdb/index"
+	"github.com/prometheus/tsdb/labels"
+)
+
+type labelValues map[string]struct{}
+
+func (lv labelValues) add(value string) {
+	lv[value] = struct{}{}
+}
+
+func (lv labelValues) get(set *[]string) {
```

Comment:
why not return?

Reply:
it reduces allocations, see caller
https://github.com/improbable-eng/thanos/pull/529/files#diff-059ebb3a930d53936a164ef53a79d7f0R244

Reply:
prevent allocations for copying slice's header on each invocation: https://blog.golang.org/go-slices-usage-and-internals

Reply:
kk